File: src/masquerade/parsers/mp4.py
"""MP4 structure verification using ISO BMFF box alignment.

The function reads the MP4 file structure, verifying that boxes (atoms)
are correctly aligned and sized according to the MP4 specification.

Specs found here: https://raw.githubusercontent.com/OpenAnsible/rust-mp4/master/docs/ISO_IEC_14496-14_2003-11-15.pdf
"""

from pathlib import Path

from masquerade.exceptions import InvalidStructureError
from masquerade.results import CheckResult
from masquerade.utils.bmff import iter_top_level_boxes, valid_ftyp

# One brand must be present in the ftyp box to consider it a valid MP4.
REQUIRED_MP4_BRANDS = {
    "mp41",  # MP4 version 1 (ISO/IEC 14496-1:2001)
    "mp42",  # MP4 version 2 (ISO/IEC 14496-14:2003)
}


def check(path: str | Path, *, max_boxes: int = 50) -> CheckResult:
    """Verify MP4 file structure by parsing ISO BMFF boxes.

    :param path: File path
    :param max_boxes: Maximum top-level boxes to scan
    :return: Structured MP4 validation outcome.
    """
    with Path(path).open("rb") as f:
        try:
            boxes = list(iter_top_level_boxes(f, max_boxes))
        except InvalidStructureError as exc:
            return CheckResult.rejected(
                "invalid_box_structure", f"Invalid MP4 box structure: {exc}"
            )

        if not boxes or boxes[0].type != "ftyp":
            return CheckResult.rejected(
                "missing_ftyp", "MP4 first box is not the required ftyp box"
            )

        if not valid_ftyp(f, boxes[0], REQUIRED_MP4_BRANDS):
            return CheckResult.rejected(
                "incompatible_brand", "MP4 ftyp box has no supported MP4 brand"
            )

        types = {box.type for box in boxes}
        missing = {"moov", "mdat"} - types
        if missing:
            missing_boxes = ", ".join(sorted(missing))
            return CheckResult.rejected(
                "missing_required_box",
                f"MP4 required boxes are missing: {missing_boxes}",
            )
        return CheckResult.accepted()

File: crates/masquerade/src/parsers/mp4.rs
//! MP4 structure verification using ISO BMFF box alignment.
//!
//! Spec: <https://raw.githubusercontent.com/OpenAnsible/rust-mp4/master/docs/ISO_IEC_14496-14_2003-11-15.pdf>

use std::io::{self, Read, Seek};

use crate::{CheckResult, ValidationLimits, containers::bmff};

const BRANDS: [[u8; 4]; 2] = [*b"mp41", *b"mp42"];
const REQUIRED_BOXES: [[u8; 4]; 2] = [*b"moov", *b"mdat"];
const SPEC: bmff::BmffSpec = bmff::BmffSpec::new("MP4", &BRANDS, &REQUIRED_BOXES);

pub(crate) fn signature<R: Read + Seek + ?Sized>(
    reader: &mut R,
    limits: &ValidationLimits,
) -> io::Result<bool> {
    SPEC.has_signature(reader, limits)
}

pub(crate) fn validate_structure<R: Read + Seek + ?Sized>(
    reader: &mut R,
    limits: &ValidationLimits,
) -> io::Result<CheckResult> {
    SPEC.validate_structure(reader, limits)
}
